File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <vector>
#include <queue>
#include <iostream>
using namespace std;

bool canGo(int n, int m, int x, int y) {
    if(x < 0 || x > n || y < 0 || y > m) return false;
    else return true;
}

int solution(vector<vector<int> > maps)
{
    queue<pair<int, int>> q;
    vector<vector<bool>> visited(maps.size(), vector<bool>(maps[0].size(), false));
    vector<vector<int>> dist(maps.size(), vector<int>(maps[0].size(), 0));
    int n = maps.size()-1;
    int m = maps[0].size()-1;
    int posX[4] = {0, -1, 0, +1};
    int posY[4] = {-1, 0, +1, 0};
    visited[0][0] = true;
    q.push(make_pair(0, 0));
    dist[0][0] = 1;
    while(!q.empty()) {
        int curX = q.front().first;
        int curY = q.front().second;
        q.pop();
        for(int i=0; i<4; i++) {
            int nextX = curX + posX[i];
            int nextY = curY + posY[i];
            if(!canGo(n, m, nextX, nextY)) continue;   // map의 범위를 벗어나는 곳
            else if(maps[nextX][nextY] == 0) continue;   // map에서 막혀있는 곳
            q.push(make_pair(nextX, nextY));
            maps[nextX][nextY] = 0;
            dist[nextX][nextY] = dist[curX][curY] + 1;
        }
    }
    if(dist[n][m] == 0) return -1;
    else return dist[n][m];
}

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Pos {
    int x;
    int y;
    int dist;
    
    public Pos(int x, int y, int dist) {
        this.x = x;
        this.y = y;
        this.dist = dist;
    }
}

class Solution {
    public int solution(int[][] maps) {
        int answer = 0;
        answer = bfs(0, 0, maps);
        return answer;
    }
    
    public boolean canGo(int x, int y, int row, int col) {
        if(x < 0 || x >= row || y < 0 || y >= col) return false;
        return true;
    }
    
    public int bfs(int x, int y, int[][] maps) {
        int row = maps.length;
        int col = maps[0].length;
        int[] xDir = {-1, 1, 0, 0};
        int[] yDir = {0, 0, -1, 1};
        boolean[][] visited = new boolean[row][col];
        
        Queue<Pos> q = new LinkedList<>();
        q.offer(new Pos(x, y, 1));
        visited[x][y] = true;
        
        while(!q.isEmpty()) {
            Pos cur = q.poll();
            int curX = cur.x;
            int curY = cur.y;
            int curDist = cur.dist;
            
            if(curX == row-1 && curY == col-1) {
                return curDist;
            }
            
            for(int i=0; i<4; i++) {
                int nextX = curX + xDir[i];
                int nextY = curY + yDir[i];
                if(canGo(nextX, nextY, row, col) && !visited[nextX][nextY] && maps[nextX][nextY] == 1) {
                    visited[nextX][nextY] = true;
                    q.offer(new Pos(nextX, nextY, curDist + 1));
                }
            }
        }
        return -1;
    }
}